kedro/io/kedro_data_catalog.py

```diff
@@ -0,0 +1,332 @@
+        self, ds_name: str, version: Version | None = None, suggest: bool = True
+    ) -> AbstractDataset:
+        ds_config = self._config_resolver.resolve_dataset_pattern(ds_name)
+
+        if ds_name not in self._datasets and ds_config is not None:
+            self._add_from_config(ds_name, ds_config)
+
+        dataset = self._datasets.get(ds_name, None)
+
+        if dataset is None:
+            error_msg = f"Dataset '{ds_name}' not found in the catalog"
+            # Flag to turn on/off fuzzy-matching which can be time consuming and
+            # slow down plugins like `kedro-viz`
+            if suggest:
+                matches = difflib.get_close_matches(ds_name, self._datasets.keys())
+                if matches:
+                    suggestions = ", ".join(matches)
+                    error_msg += f" - did you mean one of these instead: {suggestions}"
+            raise DatasetNotFoundError(error_msg)
+
+        if version and isinstance(dataset, AbstractVersionedDataset):
+            # we only want to return a similar-looking dataset,
+            # not modify the one stored in the current catalog
+            dataset = dataset._copy(_version=version)
+
+        return dataset
+
+    def _get_dataset(
+        self, dataset_name: str, version: Version | None = None, suggest: bool = True
+    ) -> AbstractDataset:
+        # TODO: remove when removing old catalog
+        return self.get_dataset(dataset_name, version, suggest)
+
+    def add(
+        self, ds_name: str, dataset: AbstractDataset, replace: bool = False
+    ) -> None:
+        """Adds a new ``AbstractDataset`` object to the ``KedroDataCatalog``."""
+        if ds_name in self._datasets:
+            if replace:
+                self._logger.warning("Replacing dataset '%s'", ds_name)
+            else:
+                raise DatasetAlreadyExistsError(
+                    f"Dataset '{ds_name}' has already been registered"
+                )
+        self._datasets[ds_name] = dataset
+
+    def list(self, regex_search: str | None = None) -> list[str]:
+        """
+        List of all dataset names registered in the catalog.
+        This can be filtered by providing an optional regular expression
+        which will only return matching keys.
+        """
+
+        if regex_search is None:
+            return list(self._datasets.keys())
+
+        if not regex_search.strip():
+            self._logger.warning("The empty string will not match any datasets")
+            return []
+
+        try:
+            pattern = re.compile(regex_search, flags=re.IGNORECASE)
+        except re.error as exc:
+            raise SyntaxError(
+                f"Invalid regular expression provided: '{regex_search}'"
+            ) from exc
+        return [ds_name for ds_name in self._datasets if pattern.search(ds_name)]
+
+    def save(self, name: str, data: Any) -> None:
+        """Save data to a registered dataset."""
+        dataset = self.get_dataset(name)
```

Comment:
```suggestion
        dataset = self[name]
```
Maybe?

Reply:
I think we should keep it as it is in case someone is trying to save an unregistered dataset. `get_dataset()` will resolve it in case it's a pattern or validate that the dataset is not in the catalog.